### PR TITLE
fix: use manifest path for xargo.toml if provided

### PR DESCRIPTION
**Pull request by priyasiddharth**
Fix for https://github.com/japaric/xargo/issues/215

### CODE REVIEW

File: src/cargo.rs [hunk 1]
```diff
@@ -250,14 +250,21 @@ impl Root {
...
         XargoMode::Build => "Cargo.toml",
         XargoMode::Check => "Xargo.toml"
     };
-    let cd = env::current_dir().chain_err(|| "couldn't get the current directory")?;
 
+    let cd = match path {
+        None => env::current_dir().chain_err(|| "couldn't get the current directory")?,
+        Some(p) => {
+            let mut pb = PathBuf::from(p);
+            pb.pop();
```
Comment (RalfJung) [inline src/cargo.rs:264]: Why do you need to pop once?

**Comment by RalfJung** [comment]
If possible it'd be great to also have a test for this. :) Our test suite is a bit fragile though, let me know if you run into trouble with it.

File: src/cargo.rs (hunk 1 above)
Comment (priyasiddharth) [reply src/cargo.rs:264]: pb is the path to `Cargo.toml`, popping gives the directory.

**Comment by priyasiddharth** [comment]
> If possible it'd be great to also have a test for this. :) Our test suite is a bit fragile though, let me know if you run into trouble with it.

Will give it a go... thanks

File: src/cargo.rs (hunk 1 above)
Comment (RalfJung) [reply src/cargo.rs:264]: `pb` comes from `args.root()` which you documented to be the root *directory*... something seems off here?

File: src/cargo.rs [hunk 2]
```diff
@@ -250,14 +250,21 @@ impl Root {
     }
 }
 
-pub fn root(mode: XargoMode) -> Result<Option<Root>> {
+pub fn root(mode: XargoMode, path: Option<&str>) -> Result<Option<Root>> {
```
Comment (RalfJung) [inline src/cargo.rs:253]: Maybe call this  `root` or `root_dir` or so, for consistency with `Args`.

File: src/cli.rs [hunk 3]
```diff
@@ -66,6 +74,7 @@ pub fn args() -> Args {
         all: all,
         subcommand: sc,
         target: target,
-        message_format: message_format
+        message_format: message_format,
+        root: root,
```
Comment (RalfJung) [inline src/cli.rs:78]:
```suggestion
        root,
```
Feel free to also do this for the other fields; this code likely predates that Rust feature.^^

File: src/cargo.rs [hunk 4]
```diff
@@ -250,14 +250,21 @@ impl Root {
...
         XargoMode::Build => "Cargo.toml",
         XargoMode::Check => "Xargo.toml"
     };
-    let cd = env::current_dir().chain_err(|| "couldn't get the current directory")?;
 
+    let cd = match manifest_path {
+        None => env::current_dir().chain_err(|| "couldn't get the current directory")?,
+        Some(p) => {
+            let mut pb = PathBuf::from(p);
+            pb.pop();
```
Comment (RalfJung) [inline src/cargo.rs:264]:
```suggestion
            pb.pop(); // strip filename, keep directory containing Cargo.toml
```

File: tests/smoke.rs [hunk 5]
```diff
@@ -227,17 +227,30 @@ impl Project {
...
     /// Calls `xargo build` and collects STDERR
     fn build_and_get_stderr(&self, target: Option<&str>) -> Result<String> {
+        self.build_from_workdir_and_get_stderr(target, &self.td.path())
+    }
+
+    fn build_from_workdir_and_get_stderr(&self, target: Option<&str>, working_dir: &Path) -> Result<String> {
         let mut cmd = xargo()?;
         cmd.arg("build");
 
+        cmd.args(&["--manifest-path", &self.td.path().join("Cargo.toml").to_str().unwrap()]);
```
Comment (RalfJung) [inline tests/smoke.rs:246]: Doesn't this mean that now *all* tests run with `--manifest-path`?

File: tests/smoke.rs [hunk 6]
```diff
@@ -445,6 +458,32 @@ fn dependencies() {
     run!()
 }
 
+/// Test building a dependency specified as `dependencies` in Xargo.toml
+#[cfg(feature = "dev")]
+#[test]
+fn dependencies_from_working_dir() {
+    fn run() -> Result<()> {
+        // need this exact target name to get the right gcc flags
+        const TARGET: &'static str = "thumbv6m-none-eabi";
```
Comment (RalfJung) [inline tests/smoke.rs:467]: Each test needs to have a different `TARGET`, to avoid using the same output directories. But the good part is that you can replace `none` by whatever you want and the targets still work -- see what the other tests are doing.

File: tests/smoke.rs [hunk 7]
```diff
@@ -445,6 +458,32 @@ fn dependencies() {
     run!()
 }
 
+/// Test building a dependency specified as `dependencies` in Xargo.toml
```
Comment (RalfJung) [inline tests/smoke.rs:461]: I don't understand this comment. I thought this is about `--manifest-path`?

File: tests/smoke.rs [hunk 8]
```diff
@@ -445,6 +458,32 @@ fn dependencies() {
...
+        project.xargo_toml(
+            r#"
+[dependencies.alloc]
+"#,
+        )?;
+        let td = TempDir::new("xargo_workdir").chain_err(|| "couldn't create a temporary directory")?;
+
+        project.build_from_workdir(TARGET, &td.path())?;
+        assert!(exists("core", TARGET)?);
+        assert!(exists("alloc", TARGET)?);
```
Comment (RalfJung) [inline tests/smoke.rs:479]: What does this have to do with `--manifest-path`?

File: tests/smoke.rs (hunk 6 above)
Comment (priyasiddharth) [reply tests/smoke.rs:467]: Ack.

File: tests/smoke.rs (hunk 5 above)
Comment (priyasiddharth) [reply tests/smoke.rs:246]: good catch! i'll fix this

File: tests/smoke.rs (hunk 8 above)
Comment (priyasiddharth) [reply tests/smoke.rs:479]: AFAICT there are no targetted tests for cmd line args for me to emulate. I was thinking that if the project built successfully, it meant that the cmdline was being processed correctly. 

If there is an easier way - I would be interested to know.

Comment (RalfJung) [reply tests/smoke.rs:479]: Well, what happens in this test without the rest of your PR?

Comment (RalfJung) [reply tests/smoke.rs:479]: I always delete `~/.xargo` before re-running xargo tests since I think it will store some state there and you want to be sure to get a fresh run.

File: src/cargo.rs (hunk 2 above)
Comment (priyasiddharth) [reply src/cargo.rs]: fixed

File: src/cli.rs (hunk 3 above)
Comment (priyasiddharth) [reply src/cli.rs]: fixed

File: tests/smoke.rs (hunk 5 above)
Comment (priyasiddharth) [reply tests/smoke.rs:256]: fixed

Comment (priyasiddharth) [reply tests/smoke.rs:256]: fixed.

File: tests/smoke.rs (hunk 8 above)
Comment (priyasiddharth) [reply tests/smoke.rs]: I've added  a `build_outside_project_dir` test which minimally tests the change.

File: tests/smoke.rs (hunk 7 above)
Comment (priyasiddharth) [reply tests/smoke.rs]: fixed

File: tests/smoke.rs [hunk 9]
```diff
@@ -445,6 +468,30 @@ fn dependencies() {
...
+        // need this exact target name to get the right gcc flags
+        const TARGET: &'static str = "thumbv6m-workdir-eabi";
+
+        let project = Project::new(TARGET)?;
+        project.xargo_toml(
+            r#"
+[dependencies.alloc]
+"#,
+        )?;
+        let td = TempDir::new("xargo_workdir").chain_err(|| "couldn't create a temporary directory")?;
```
Comment (RalfJung) [inline tests/smoke.rs:485]: So crucially, this is a `TempDir` that is *separate* from the `TempDir` inside `project`, right? If so, please add a comment explaining that.

Comment (priyasiddharth) [reply tests/smoke.rs]: correct.fixed.

**Comment by RalfJung** [comment]
All right, looking good -- thanks @priyasiddharth!

bors r+

**Comment by priyasiddharth** [comment]
> All right, looking good -- thanks @priyasiddharth!
> 
> bors r+

Thanks @RalfJung for the timely review!
